[InstCombine][asan] Don't speculate loads before `select ptr`

vitalybuka opened the pull request:
Even if memory is valid from `llvm` point of view,
e.g. local alloca, sanitizers have API for user
specific memory annotations.

These annotations can be used to track size of the
local object, e.g. inline vectors may prevent
accesses beyond the current vector size.

So valid programs should not access those parts of
alloca before checking preconditions.

Fixes #100639.

vitalybuka committed:
[InstCombine][asan] Don't speculate loads before `select ptr`

Even if memory is valid from `llvm` point of view,
e.g. local alloca, sanitizers have API for user
specific memory annotations.

This annotations can be used to track size of the
local object, e.g. inline vector like may prevent
accessed beyond the current vector size.

So valid programs should not access those parts of
alloca before checking preconditions.

Fixes #100639.

Pull Request: https://github.com/llvm/llvm-project/pull/100773

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineLoadStoreAlloca.cpp:
```diff
@@ -1042,8 +1042,8 @@ Instruction *InstCombinerImpl::visitLoadInst(LoadInst &LI) {
-  // loads.  Most of them do apply for unordered atomics.
-  if (!LI.isUnordered()) return nullptr;
+  // loads and sanitizers.  Most of them do apply for unordered atomics.
+  if (mustSuppressSpeculation(LI)) return nullptr;
```
Can we move this check to where it is relevant, i.e. the select fold in particular?

Also, should this check be inside isSafeToLoadUnconditionally maybe?

vitalybuka replied:
Done.

vitalybuka committed:
[NFC][Load] Make `ScanFrom` required parameters

In #100773 we will go conservative for sanitizers,
so it's better to pinpoint location consciously.

Pull Request: https://github.com/llvm/llvm-project/pull/100789

vitalybuka committed:
[NFC][Load] Find better place for `mustSuppressSpeculation`

And extract `suppressSpeculativeLoadForSanitizers`.

Pull Request: https://github.com/llvm/llvm-project/pull/100794

vitalybuka committed:
[NFC][InstCombine][SROA][Asan] Precommit test affected by #100773

Some optimization need to be undone with
sanitizers by #100773.

Pull Request: https://github.com/llvm/llvm-project/pull/100844

nikic reviewed (approved):
LGTM

vitalybuka committed:
rebase

Created using spr 1.3.4
